### PR TITLE
One shot export [Image classification]

### PR DESCRIPTION
**Pull request by rahul-tuli**
This PR adds support to apply a recipe in `one-shot` manner while exporting Image classification models

Test plan: Automated test to start with a vanilla(non-quantized) `resnet50`, export checkpoint in one-shot manner with a quantization recipe and then check the onnx graph for quantized nodes

Note the failing test should pass after #1122 lands

### CODE REVIEW

File: src/sparseml/pytorch/image_classification/export.py [hunk 1]
```diff
@@ -290,6 +297,13 @@ class strings. Expected to be a path to a
     show_default=True,
     help="number of classes for model; must be set if a dataset is not provided",
 )
+@click.option(
+    "--one-shot",
+    "--one_shot",
+    is_flag=True,
+    show_default=True,
```
Comment (corey-nm) [inline src/sparseml/pytorch/image_classification/export.py:304]: missing default value?

File: src/sparseml/pytorch/image_classification/export.py [hunk 2]
```diff
@@ -366,9 +381,18 @@ def main(
...
         labels_to_class_mapping = helpers.label_to_class_mapping_from_dataset(dataset)
 
-    if recipe is not None:
-        ScheduledModifierManager.from_yaml(recipe).apply_structure(model)
-        load_model(checkpoint_path, model, strict=True)
+    # training recipe must be a part of checkpoint
+    load_model(checkpoint_path, model, strict=True)
+
+    if one_shot:
+        if not recipe:
```
Comment (corey-nm) [inline src/sparseml/pytorch/image_classification/export.py:388]: Do we need to cover the case where `recipe==True` and `one_shot==False`?

File: tests/sparseml/pytorch/image_classification/test_export.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+import pytest
+import torch
+
+from click.testing import CliRunner
+from sparseml.pytorch.image_classification.export import main
+from sparseml.pytorch.models import resnet50
+
+
+@pytest.fixture()
+def temp_dir():
```
Comment (corey-nm) [inline tests/sparseml/pytorch/image_classification/test_export.py:28]: you can just use the built in `tmp_path` fixture instead of this 😀

File: tests/sparseml/pytorch/image_classification/test_export.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+    # exported file exists
+    assert os.path.exists(expected_onnx_path)
+
+    # exported file is valid onnx
+    model = onnx.load(expected_onnx_path)
+    onnx.checker.check_model(model)
+
+    # check onnx model is quantized
+    assert any(node.op_type == "QuantizeLinear" for node in model.graph.node)
```
Comment (corey-nm) [inline tests/sparseml/pytorch/image_classification/test_export.py:90]: nice test!! 🎉

File: src/sparseml/pytorch/image_classification/export.py (hunk 1 above)
Comment (rahul-tuli) [reply src/sparseml/pytorch/image_classification/export.py:304]: Flag always has a default value of `False`

File: src/sparseml/pytorch/image_classification/export.py (hunk 2 above)
Comment (rahul-tuli) [reply src/sparseml/pytorch/image_classification/export.py:388]: Good catch! Updating.

File: tests/sparseml/pytorch/image_classification/test_export.py (hunk 3 above)
Comment (rahul-tuli) [reply tests/sparseml/pytorch/image_classification/test_export.py:28]: Thanks!

File: src/sparseml/pytorch/image_classification/export.py [hunk 5]
```diff
@@ -368,7 +386,13 @@ def main(
 
     if recipe is not None:
         ScheduledModifierManager.from_yaml(recipe).apply_structure(model)
+
+    if checkpoint_path:
         load_model(checkpoint_path, model, strict=True)
+
+    if one_shot is not None:
```
[the remark this answers is not in the document]
Comment (rahul-tuli) [reply src/sparseml/pytorch/image_classification/export.py:393]: Yes, both can exist, recipe is generally a training recipe, at some point in our codebase we moved to saving the training recipe in the checkpoint itself like `{'state_dict': ..., 'recipe': ...}` the training recipe being provided separately is to support older flows;

This one shot export flow, is meant so that someone can quickly test how the model will look like after application of another recipe.

for example, the original training recipe could quantize some parts of the graph; later if a user wants to apply another recipe on top that quantizes different layers or weights, they could pass in the `new` recipe with the one shot argument; if they saved the checkpoint without training recipe they could pass in the `old` training recipe with the `--recipe` argument;
like the following:
```
--recipe old_training_recipe.md --one-shot new_one_shot_recipe.md
```

What would happen here essentially is that the training recipe will be used to apply structure to the model then load weights, after which the one shot recipe will be applied and the model will be exported. Does this make sense?

**Review by corey-nm — APPROVED** [review]
🎉

File: src/sparseml/pytorch/image_classification/export.py [hunk 6]
```diff
@@ -368,7 +386,13 @@ def main(
 
     if recipe is not None:
         ScheduledModifierManager.from_yaml(recipe).apply_structure(model)
+
+    if checkpoint_path:
```
Comment (bfineran) [inline src/sparseml/pytorch/image_classification/export.py:390]: what is the reason for this change? could this potentially fail before?

File: tests/sparseml/pytorch/image_classification/test_export.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+def quant_recipe_path():
+    """
+    return: path to a resnet50 recipe with quantization
+    """
+    recipe_path = "./resnet_test_recipe.md"
+    recipe_path = os.path.abspath(recipe_path)
+    yield recipe_path
+
+
+def test_export_one_shot(non_quant_resnet_checkpoint, quant_recipe_path, temp_dir):
```
Comment (bfineran) [inline tests/sparseml/pytorch/image_classification/test_export.py:57]: we should be checking pruning here as well, especially since you're using a pruning recipe. (use the `tensor_sparsity` helper, see comments in the recipe template PR)

File: tests/sparseml/pytorch/image_classification/resnet_test_recipe.md [hunk 8]
```diff
@@ -0,0 +1,207 @@
+<!--
```
Comment (bfineran) [inline tests/sparseml/pytorch/image_classification/resnet_test_recipe.md:1]: this file is unnecessary IMO - copied from a zoo recipe contains bloat such as non-relevant modifiers and hyper parameters as well as the markdown.

We should be able to just move this into a constant string in python

```yaml
_TEST_RECIPE_STR = """
modifiers:
  - !ACDCPruningModifier
    compression_sparsity: 0.9
    start_epoch: 0.0
    end_epoch: 1.0
    global_sparsity: True 
    update_frequency: 1.0
    params: __ALL_PRUNABLE__

  - !QuantizationModifier
    start_epoch: 1.0
    disable_quantization_observer_epoch: 2.0
    freeze_bn_stats_epoch: 2.0
"""
```
should suffice

**Review by bfineran — CHANGES_REQUESTED** [review]
few quick changes to test then LGTM

File: src/sparseml/pytorch/image_classification/export.py (hunk 6 above)
Comment (rahul-tuli) [reply src/sparseml/pytorch/image_classification/export.py:390]: Before, the code wasn't loading in checkpoint weights if the recipe was not provided which was a bug; this fixes that

File: tests/sparseml/pytorch/image_classification/test_export.py (hunk 7 above)
Comment (rahul-tuli) [reply tests/sparseml/pytorch/image_classification/test_export.py:57]: Tensor sparsity takes in a torch tensor, I used ModelAnalysis tool in upcoming commit to assert sparsity

File: tests/sparseml/pytorch/image_classification/resnet_test_recipe.md (hunk 8 above)
Comment (rahul-tuli) [reply tests/sparseml/pytorch/image_classification/resnet_test_recipe.md:1]: Thanks for the minimal recipe, I would argue we should still keep this as a test file, point being a recipe file path is expected by the export script, we can always write the recipe to a temporary file on the fly during testing, let me know your thoughts on that

File: tests/sparseml/pytorch/image_classification/test_export.py (hunk 4 above)
Comment (rahul-tuli) [reply tests/sparseml/pytorch/image_classification/test_export.py:88]: Thanks @corey-nm!

**Review by corey-nm — DISMISSED** [review]
lgtm

File: tests/sparseml/pytorch/image_classification/test_export.py [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+
+    # check onnx model is quantized
+    assert any(node.op_type == "QuantizeLinear" for node in model.graph.node)
+
+    # check onnx model is sparse
+    model_analysis = ModelAnalysis.from_onnx(onnx_file_path=expected_onnx_path)
+    block = model_analysis.operation_summary.ops.block_structure
+
+    assert block["single"].sparsity > 0.99
+    assert block["block4"].sparsity > 0.95
```
Comment (bfineran) [inline tests/sparseml/pytorch/image_classification/test_export.py:95]: this doesn't seem right to me - the recipe is only targeting 0.9% sparsity, which means the single and block sparsity should be much lower. additionally since the recipe isn't targeting block sparsity explicitly, we shouldn't be testing for it. any block sparsity is a side effect

**Review by bfineran — CHANGES_REQUESTED** [review]

File: tests/sparseml/pytorch/image_classification/test_export.py (hunk 9 above)
Comment (rahul-tuli) [reply tests/sparseml/pytorch/image_classification/test_export.py]: Updated!

**Review by KSGulin — APPROVED** [review]

File: tests/sparseml/pytorch/image_classification/test_export.py [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+    onnx.checker.check_model(model)
+
+    # check onnx model is quantized
+    assert any(node.op_type == "QuantizeLinear" for node in model.graph.node)
+
+    # check onnx model is sparse
+    model_analysis = ModelAnalysis.from_onnx(onnx_file_path=expected_onnx_path)
+    block = model_analysis.operation_summary.ops.block_structure
+
+    assert block["single"].sparsity > 0.90
```
Comment (bfineran) [inline tests/sparseml/pytorch/image_classification/test_export.py:94]: this still doesn't quite add up (doesn't check the previous issue that the model was way too sparse, likely all 0s) - sparsity shouldn't be greater than 0.9 since we are targeting 0.9. should test that we are within a margin of error.

**Review by bfineran — CHANGES_REQUESTED** [review]
another testing comment, also please look into the failing tests

File: tests/sparseml/pytorch/image_classification/test_export.py (hunk 10 above)
Comment (rahul-tuli) [reply tests/sparseml/pytorch/image_classification/test_export.py:94]: What's an acceptable margin of error?

**Review by bfineran — APPROVED** [review]

**Comment by bfineran** [comment]
bumping this

File: tests/sparseml/pytorch/image_classification/test_export.py [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+
+
+def test_export_one_shot(non_quant_resnet_checkpoint, quant_recipe_path, temp_dir):
+    runner = CliRunner()
+    arch_key = "resnet50"
+    result = runner.invoke(
+        main,
+        [
+            "--arch-key",
+            "resnet50",
```
Comment (bfineran) [inline tests/sparseml/pytorch/image_classification/test_export.py:64]: let's change to resnet18 or mobilenet

**Review by bfineran — APPROVED** [review]

**Review by corey-nm — APPROVED** [review]
🔥
